[GSoD]  Added tutorials directory on gRPC-Gateway (#1829)
grpc-ecosystem/grpc-gateway: docs/docs/tutorials/Generating stubs/using_buf.md

---
layout: default
title: Generating stubs using buf
parent: Generating stubs
grand_parent: Tutorials
nav_order: 1
---

## Generating stubs using buf

Buf is configured through a `buf.yaml` file that should be checked in to the root of your repository. Buf will automatically read this file if present. Configuration can also be provided via the command-line flag `--config`, which accepts a path to a `.json` or `.yaml` file, or direct JSON or YAML data.

All Buf operations that use your local `.proto` files as input rely on a valid build configuration. This configuration tells Buf where to search for `.proto` files, and how to handle imports. As opposed to protoc, where all `.proto` files are manually specified on the command-line, buf operates by recursively discovering all `.proto` files under configuration and building them.

> The following is an example of all configuration options for the build.

Reviewer: ```suggestion
The following is an example of a valid configuration, assuming you have your `.proto` files rooted in the `proto` folder relative to the root of your repository.
```

```yml
version: v1beta1
build:
  roots:
    - proto

>     - vendor/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis

Reviewer: I think for this part of the tutorial we don't include the `http.proto` annotations, we just use vanilla protobuf so to speak, so this line is not necessary.

  excludes:

>     - proto/foo/bar

Reviewer: Lets remove this, it's not related to generation.

```

> To generate stubs for for multiple languages. Create the file `buf.gen.yaml` at the root of the repository:

Reviewer: Lets focus on the simple case of generating for a single language, Go.
```suggestion
To generate type and gRPC stubs for Go, create the file `buf.gen.yaml` at the root of the repository:
```

```yml
version: v1beta1

> plugins:
>   - name: java
>     out: java
>   - name: cpp
>     out: cpp

Reviewer: ```suggestion
plugins:
  - name: go
    out: proto
    opt: paths=source_relative
  - name: go-grpc
    out: proto
    opt: paths=source_relative
```

```

> 

Reviewer: ```suggestion

We use the `go` and `go-grpc` plugins to generate Go types and gRPC service definitions. We're outputting the generated files relative to the `proto` folder, and we're using the `paths=source_relative` option, which means that the generated files will appear in the same directory as the source `.proto` file.
```

Then run

```sh

> buf generate --file ./proto/example.proto

Reviewer: The recommended way to run this is without specifying anything, so it generates stubs for all files.
```suggestion
buf generate
```

> ```

Reviewer: ```suggestion
''' (replace with backticks, I just want us to add the below)
This will have generated a `*.pb.go` and a `*_grpc.pb.go` file for each protobuf package in our `proto` file hierarchy.
```
